// yeoman-generator/dist/util/storage.js
import assert from 'node:assert';
import { cloneDeep, get, merge, set, defaults as setDefaults } from 'lodash-es';
import sortKeys from 'sort-keys';
/**
 * Proxy handler for Storage
 */
const proxyHandler = {
    get(storage, property, _receiver) {
        return storage.get(property, { raw: true });
    },
    set(storage, property, value, _receiver) {
        if (typeof property === 'string') {
            storage.set(property, value);
            return true;
        }
        return false;
    },
    ownKeys(storage) {
        return Reflect.ownKeys(storage._store);
    },
    has(storage, prop) {
        return storage.get(prop, { raw: true }) !== undefined;
    },
    getOwnPropertyDescriptor(storage, key) {
        return {
            get: () => this.get(storage, key, null),
            enumerable: true,
            configurable: true,
        };
    },
};
/**
 * Storage instances handle a json file where Generator authors can store data.
 *
 * The `Generator` class instantiate the storage named `config` by default.
 *
 * @constructor
 * @param name     The name of the new storage (this is a namespace)
 * @param fs  A mem-fs editor instance
 * @param configPath The filepath used as a storage.
 *
 * @example
 * class extend Generator {
 *   writing: function() {
 *     this.config.set('coffeescript', false);
 *   }
 * }
 */
class Storage {
    path;
    name;
    fs;
    indent;
    lodashPath;
    disableCache;
    disableCacheByFile;
    sorted;
    existed;
    _cachedStore;
    transform;
    constructor(name, fs, configPath, options = {}) {
        let editor;
        let actualName;
        let actualConfigPath;
        let actualOptions = options;
        if (typeof name === 'string') {
            actualName = name;
        }
        else if (typeof name === 'object') {
            editor = name;
        }
        if (typeof fs === 'string') {
            actualConfigPath = fs;
        }
        else {
            editor = fs;
        }
        if (typeof configPath === 'string') {
            actualConfigPath = configPath;
        }
        else if (typeof configPath === 'object') {
            actualOptions = configPath;
        }
        if (!editor) {
            throw new Error(`Check parameters`);
        }
        assert.ok(actualConfigPath, 'A config filepath is required to create a storage');
        this.path = actualConfigPath;
        this.name = actualName ?? actualOptions.name;
        this.fs = editor;
        this.indent = 2;
        this.lodashPath = actualOptions.lodashPath ?? false;
        this.disableCache = actualOptions.disableCache ?? false;
        this.disableCacheByFile = actualOptions.disableCacheByFile ?? false;
        this.sorted = actualOptions.sorted ?? false;
        this.transform = actualOptions.transform;
        this.existed = Object.keys(this._store).length > 0;
        this.fs.store.on('change', filename => {
            // At mem-fs 1.1.3 filename is not passed to the event.
            if (this.disableCacheByFile || (filename && filename !== this.path)) {
                return;
            }
            delete this._cachedStore;
        });
    }
    /**
     * @protected
     * @return the store content
     */
    readContent() {
        const content = this.fs.readJSON(this.path, {});
        if (!content || typeof content !== 'object' || Array.isArray(content)) {
            throw new Error(`${this.path} is not a valid Storage`);
        }
        return content;
    }
    /**
     * @protected
     * @return the store content
     */
    writeContent(fullStore) {
        return this.fs.writeJSON(this.path, fullStore, undefined, this.indent);
    }
    /**
     * Return the current store as JSON object
     * @return the store content
     * @private
     */
    get _store() {
        const store = this._cachedStore ?? this.readContent();
        if (!this.disableCache) {
            this._cachedStore = store;
        }
        if (!this.name) {
            return store;
        }
        return ((this.lodashPath ? get(store, this.name) : store[this.name]) ?? {});
    }
    /**
     * Persist a configuration to disk
     * @param val - current configuration values
     * @private
     */
    _persist(value) {
        if (this.sorted) {
            value = sortKeys(value, { deep: true });
        }
        let fullStore;
        if (this.name) {
            fullStore = this.readContent();
            if (this.lodashPath) {
                set(fullStore, this.name, value);
            }
            else {
                fullStore[this.name] = value;
            }
        }
        else {
            fullStore = value;
        }
        this.writeContent(fullStore);
    }
    /**
     * Save a new object of values
     */
    save() {
        this._persist(this._store);
    }
    /**
     * Get a stored value
     * @param key  The key under which the value is stored.
     * @return The stored value. Any JSON valid type could be returned
     */
    get(key, { raw } = {}) {
        if (raw) {
            return this._store[key];
        }
        return this.transformedStore()[key];
    }
    /**
     * Get a stored value from a lodash path
     * @param path  The path under which the value is stored.
     * @return The stored value. Any JSON valid type could be returned
     */
    getPath(path) {
        return get(this.transformedStore(), path);
    }
    /**
     * Get all the stored values
     * @return key-value object
     */
    getAll() {
        return this.transformedStore({ forceClone: true });
    }
    set(key, value) {
        const store = this._store;
        let ret;
        if (typeof key === 'object') {
            ret = Object.assign(store, key);
        }
        else if (typeof key === 'string' || typeof key === 'number') {
            store[key] = value;
            ret = value;
        }
        else {
            throw new TypeError(`key not supported ${typeof key}`);
        }
        this._persist(store);
        return ret;
    }
    /**
     * Assign a lodash path to a value and schedule a save.
     * @param path  The key under which the value is stored
     * @param val  Any valid JSON type value (String, Number, Array, Object).
     * @return val  Whatever was passed in as val.
     */
    setPath(path, value) {
        assert.ok(typeof value !== 'function', "Storage value can't be a function");
        const store = this._store;
        set(store, path, value);
        this._persist(store);
        return value;
    }
    /**
     * Delete a key from the store and schedule a save.
     * @param key  The key under which the value is stored.
     */
    delete(key) {
        const store = this._store;
        delete store[key];
        this._persist(store);
    }
    /**
     * Setup the store with defaults value and schedule a save.
     * If keys already exist, the initial value is kept.
     * @param defaults  Key-value object to store.
     * @return val  Returns the merged options.
     */
    defaults(defaults) {
        assert.ok(typeof defaults === 'object', 'Storage `defaults` method only accept objects');
        const store = setDefaults({}, this._store, defaults);
        this._persist(store);
        return this.getAll();
    }
    /**
     * @param defaults  Key-value object to store.
     * @return val  Returns the merged object.
     */
    merge(source) {
        assert.ok(typeof source === 'object', 'Storage `merge` method only accept objects');
        const value = merge({}, this._store, source);
        this._persist(value);
        return this.getAll();
    }
    createStorage(path) {
        const childName = this.name ? `${this.name}.${path}` : path;
        return new Storage(childName, this.fs, this.path, { lodashPath: true, transform: this.transform });
    }
    /**
     * Creates a proxy object.
     * @return proxy.
     */
    createProxy() {
        return new Proxy(this, proxyHandler);
    }
    transformedStore({ forceClone = false } = {}) {
        if (this.transform) {
            return this.transform(cloneDeep(this._store), this.name);
        }
        if (forceClone) {
            return cloneDeep(this._store);
        }
        return this._store;
    }
}
export default Storage;
