// yeoman-generator/dist/index.js
import process from 'node:process';
import { CheckRepoActions, CleanOptions, simpleGit } from 'simple-git';
import { BaseGenerator } from './generator.js';
export { default as Storage } from './util/storage.js';
export default class Generator extends BaseGenerator {
    constructor(arguments_, options, features) {
        super(arguments_, options, features);
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-expect-error
        this._queues = {};
        // Add original queues.
        for (const queue of BaseGenerator.queues) {
            this._queues[queue] = { priorityName: queue, queueName: queue };
        }
        // Add custom queues
        if (Array.isArray(this.features.customPriorities)) {
            this.registerPriorities(this.features.customPriorities);
        }
    }
    get simpleGit() {
        return this.createSimpleGit();
    }
    createSimpleGit(options) {
        const git = simpleGit({ baseDir: this.destinationPath(), ...options }).env({
            HOME: process.env.HOME,
            PATH: process.env.PATH,
            LANG: 'C',
            LC_ALL: 'C',
        });
        git.CheckRepoActions = CheckRepoActions;
        git.CleanOptions = CleanOptions;
        return git;
    }
}
