// mem-fs-editor/dist/state.js
import fs from 'fs';
const states = {
    MODIFIED: 'modified',
    DELETED: 'deleted',
};
export const setFileState = (file, state) => {
    file.state = state;
};
export const isFileNew = (file) => {
    if (file.isNew === undefined) {
        file.isNew = !fs.existsSync(file.path);
    }
    return file.isNew;
};
export const isFileStateModified = (file) => file.state === states.MODIFIED;
export const setModifiedFileState = (file) => {
    setFileState(file, states.MODIFIED);
};
export const isFileStateDeleted = (file) => file.state === states.DELETED;
export const setDeletedFileState = (file) => {
    setFileState(file, states.DELETED);
};
export const isFilePending = (file) => isFileStateModified(file) || (isFileStateDeleted(file) && !isFileNew(file));
export const setCommittedFile = (file) => {
    file.committed = true;
};
export const isFileCommitted = (file) => Boolean(file.committed);
export const resetFileState = (file) => {
    delete file.state;
};
/**
 * Delete commit related states.
 */
export const resetFileCommitStates = (file) => {
    delete file.stateCleared;
    delete file.committed;
};
/**
 * Delete all mem-fs-editor`s related states.
 */
export const resetFile = (file) => {
    resetFileState(file);
    resetFileCommitStates(file);
    delete file.isNew;
};
export const clearFileState = (file) => {
    if (file.state) {
        file.stateCleared = file.state;
    }
    resetFileState(file);
    delete file.isNew;
};
export const hasState = (file) => Boolean(file.state);
export const hasClearedState = (file) => Boolean(file.stateCleared);
