// mem-fs-editor/dist/actions/copy-async.js
import assert from 'assert';
import fs from 'fs';
import fsPromises from 'fs/promises';
import path from 'path';
import createDebug from 'debug';
import { glob, isDynamicPattern } from 'tinyglobby';
import multimatch from 'multimatch';
import normalize from 'normalize-path';
import File from 'vinyl';
import { resolveFromPaths, getCommonPath, globify, resolveGlobOptions } from '../util.js';
import { writeInternal } from './write.js';
const debug = createDebug('mem-fs-editor:copy-async');
async function getOneFile(filepath) {
    const resolved = path.resolve(filepath);
    try {
        if ((await fsPromises.stat(resolved)).isFile()) {
            return resolved;
        }
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
    }
    catch (error) { }
    return undefined;
}
export async function copyAsync(from, to, options = {}) {
    to = path.resolve(to);
    const { noGlob } = options;
    const hasGlobOptions = Boolean(options.globOptions);
    const hasMultimatchOptions = Boolean(options.storeMatchOptions);
    assert(!noGlob || !hasGlobOptions, '`noGlob` and `globOptions` are mutually exclusive');
    assert(!noGlob || !hasMultimatchOptions, '`noGlob` and `storeMatchOptions` are mutually exclusive');
    if (typeof from === 'string') {
        const oneFile = await getOneFile(from);
        if (oneFile) {
            return copySingleAsync(this, oneFile, to, options);
        }
    }
    const { fromBasePath = getCommonPath(from) } = options;
    const resolvedFromPaths = resolveFromPaths({ from, fromBasePath });
    const hasDynamicPattern = resolvedFromPaths.some((f) => isDynamicPattern(normalize(f.from)));
    const { preferFiles } = resolveGlobOptions({
        noGlob,
        hasDynamicPattern,
        hasGlobOptions,
    });
    const storeFiles = [];
    const globResolved = [];
    for (const resolvedFromPath of resolvedFromPaths) {
        const { resolvedFrom } = resolvedFromPath;
        if (this.store.existsInMemory(resolvedFrom)) {
            storeFiles.push(resolvedFrom);
        }
        else {
            globResolved.push(resolvedFromPath);
        }
    }
    let diskFiles = [];
    if (globResolved.length > 0) {
        const patterns = globResolved.map((file) => globify(file.from)).flat();
        diskFiles = (await glob(patterns, { cwd: fromBasePath, ...options.globOptions, absolute: true, onlyFiles: true })).map((file) => path.resolve(file));
        const normalizedStoreFilePaths = this.store
            .all()
            .map((file) => file.path)
            .filter((filePath) => !diskFiles.includes(filePath))
            .map((filePath) => normalize(filePath))
            // The store may have a glob path and when we try to copy it will fail because not real file
            .filter((filePath) => !isDynamicPattern(filePath));
        multimatch(normalizedStoreFilePaths, patterns.map((p) => (path.isAbsolute(p) ? p : path.posix.join(normalize(fromBasePath), p))), options.storeMatchOptions).forEach((filePath) => {
            storeFiles.push(path.resolve(filePath));
        });
    }
    // Sanity checks: Makes sure we copy at least one file.
    assert(options.ignoreNoMatch || diskFiles.length > 0 || storeFiles.length > 0, 'Trying to copy from a source that does not exist: ' + String(from));
    // If `from` is an array, or if it contains any dynamic patterns, or if it doesn't exist, `to` must be a directory.
    const treatToAsDir = Array.isArray(from) || !preferFiles || globResolved.length > 0;
    let generateDestination = () => to;
    if (treatToAsDir) {
        assert(!this.exists(to) || fs.statSync(to).isDirectory(), 'When copying multiple files, provide a directory as destination');
        generateDestination = (filepath) => path.join(to, path.relative(fromBasePath, filepath));
    }
    await Promise.all([
        ...diskFiles.map((file) => copySingleAsync(this, file, generateDestination(file), options)),
        ...storeFiles.map((file) => copySingleAsync(this, file, generateDestination(file), options)),
    ]);
}
const defaultFileTransform = ({ destinationPath, contents }) => ({
    path: destinationPath,
    contents,
});
async function copySingleAsync(editor, from, to, options = {}) {
    from = path.resolve(from);
    debug('Copying %s to %s with %o', from, to, options);
    const file = editor.store.get(from);
    assert(file.contents, `Cannot copy empty file ${from}`);
    const { fileTransform = defaultFileTransform, transformOptions, transformData } = options;
    const transformPromise = fileTransform({
        destinationPath: path.resolve(to),
        sourcePath: from,
        contents: file.contents,
        options: transformOptions,
        data: transformData,
    });
    let contents;
    ({ path: to, contents } = await transformPromise);
    if (options.append && editor.store.existsInMemory(to)) {
        editor.append(to, contents, { create: true, ...options });
    }
    else if (File.isVinyl(file)) {
        writeInternal(editor.store, Object.assign(file.clone({ contents: false, deep: false }), {
            contents: Buffer.from(contents),
            path: to,
        }));
    }
    else {
        writeInternal(editor.store, new File({
            contents: Buffer.from(contents),
            stat: file.stat,
            path: to,
            history: [file.path],
        }));
    }
}
